### ISSUE TITLE
XML Empty tag to Empty string in the object during xml deserialization

### ISSUE THREAD

**Issue by thefalconfeat** [FasterXML/jackson-dataformat-xml#162]
I am using jackson-dataformat-xml:2.6.2

If an empty tag is available in the xml and the respective field in the object is a string, it would be nice to have jackson populate the field with empty string rather than with null value.

Ex:

``` java
@XmlAccessorType(XmlAccessType.FIELD)
@XmlType(
        name = "",
        propOrder = {"first", "last"}
)
@XmlRootElement(
        name = "name"
)
public class Name {
private String first;
private String last;
}
```

xml:

``` xml
<name>
<first>Ryan</first>
<last></last>
</name>
```

mapper creation:

```
new XmlMapper(XMLInputFactory2.newFactory(), XMLOutputFactory2.newFactory());
```

I would expect 

```
Name name = mapper.readValue(xml, Name.class)
```

name.getLast().equals("") to return true but name.getLast() ends up being null.

Any help would be appreciated.

**Comment by cowtowncoder** [comment]
Yes I can reproduce this. It is a tricky problem to solve but I added a failing unit test so that hopefully it may be resolved in future.

In the meantime a work-around would be to add explicit `setLast()` that can convert `null` into empty String (and perhaps also pre-initialize `last` as empty string in constructor.

**Comment by cowtowncoder** [comment]
I have some ideas as to how this could be implemented; but actual fix will need to wait until 2.8 unfortunately, as some API changes will be needed.

**Comment by RobinAllen** [comment]
I'm also facing what I think is the same issue.   
In my case (simplified here to just the basic problem), I'm reading xml which defines a 'Thing's state as A or B:

State A : `"<Thing><A /></Thing>"`
State B : `"<Thing><B /></Thing>"`

Both A and B are defined in XSD as being 'xsd:string' with max length 0 characters. 'Thing' is a complextype, defined as being <xsd:choice> between element A or element B. They are _intended_ to have no content; it is purely the _presence_ or _absence_ of the empty element which declares which state we are in. Sadly I don't have control over the interface to be able to change the logic.  I have to live with this XML interface.

In Java, the POJO generated from XSD has String fields 'a' and 'b'.

If I serialize from POJO to XML I have no problem, but the trouble comes when I try to deserialize again.

If I deserialize `"<Thing><A /></Thing>"`, then I want to end up with [ a="", b=null ]
If I deserialize `"<Thing><B /></Thing>"`, then I want to end up with [ a=null, b="" ]
But whichever input I deserialize, I get the same results:  [ a=null, b=null ]

So you see, I'm unable to detect _WHICH_ empty element was present, which leaves me unable to know which state it was in.

If I could submit some content inside the A / B, even just a space character, then it would deserialize properly.  But I can't _validly_ submit content inside the A / B element, because that would not be acceptable according to the XSD (max length 0 characters).

I see this as a fault, really, because dataformat-xml cannot read back in the xml which it itself has outputted.  If I serialize a Thing from POJO to XML string and then deserialize back to POJO again, I don't get back in the same object I wrote out.  (Quoting the readme page:  "_What should be guaranteed is that any XML written using this module must be readable using module as well_")

**Comment by cowtowncoder** [comment]
@RobinAllen This is indeed problematic, but consider that without mandating use of XML Schema's nillable, there really is nothing to definitely differentiate empty String from `null`.

Anyway, I do hope that with 2.8 (to be worked on next now that 2.7.0 is ready to release) we can at least ensure that empty element, for `String` valued property, will become "". And total absence of element would in that case remain `null`. And finally, maybe nullability with XML schema can be eventually supported as well, for explicit nulls.

**Comment by ihommani** [comment]
Hi @cowtowncoder,

I may be wrong but i think my need is related to this issue. 
Given the following class:

`
public class TestJacksonList {

    List<String> strings;
    List<Integer> integers;
    List<Path> paths;
    Integer integer;

    public static final ObjectMapper XML_MAPPER = new XmlMapper()
            .setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY)
            .setVisibility(PropertyAccessor.GETTER, JsonAutoDetect.Visibility.NONE);


    public static TestJacksonList parseXml(String content) throws IOException {
        try {
            TestJacksonList testJacksonList = XML_MAPPER.readValue(content, TestJacksonList.class);
            return testJacksonList;
        } catch (IOException e) {
            throw e;
        }
    }

    public List<String> getStrings() {
        return strings;
    }

    public List<Path> getPaths() {
        return paths;
    }

    public List<Integer> getIntegers() {
        return integers;
    }

    public Integer getInteger() {
        return integer;
    }
}`

I am unable to deserialize this kind of xml: 
` <xml><paths>
<path>hello</path><path>world</path><path></path>
</paths></xml>`

Yet Jackson understand empty Integer/Path tag by setting it to null.
As soon as we get into a list, this behavior is no more observed and jackson fail on a START_OBJECT. 
I would expect the list [Path(hello), Path(world), null]
I opened a thread on stackoverflow for more details and to know if I am missing smthg obvious.
http://stackoverflow.com/questions/38301909/jackson-collection-deserializer-to-accept-empty-xml-tags

Thx

**Comment by cowtowncoder** [comment]
@ihommani I wouldn't guess this is the same thing: you may want to file a separate issue since what you see may or may not be related. You can just add a reference to this issue as possibly related, but before establishing that it is better to keep potentially separate problems in separate issues.

**Comment by ihommani** [comment]
@cowtowncoder , got it. I erase my previous post and create a new issue.

**Comment by manueljordan** [comment]
Hello 

what is the status of this situation? 

I am working with `Jackson 2.8.2` and seems this situation is reflected until now in:
[Spring Rest & Jackson: Empty values are transformed in Null values in XML transformation](http://stackoverflow.com/questions/40253391/spring-rest-jackson-empty-values-are-transformed-in-null-values-in-xml-transf)

Thanks

**Comment by cowtowncoder** [comment]
@manueljordan If there is progress, there will be update here. I have not worked on this, and am unlikely to have much time in near future.

**Comment by manueljordan** [comment]
@cowtowncoder thanks by the quick reply. I understand. The important is it was reported. Yes, take your time, I know you are busy. Meanwhile I have turned around this situation in other way.

**Comment by jlous** [comment]
Any idea when the PR might be merged if approved?

**Comment by manueljordan** [comment]
What is needed to approve the PR?

**Comment by oistein** [comment]
Yeah, what is stopping this?

**Comment by AlonsoZhou** [comment]
@cowtowncoder 
Hello,what is the status of this situation?and the Which is version that the bug is fiexed?

**Comment by jhaber** [comment]
This is a pretty painful issue to run into, but after a lot of not-so-fun experimentation I came up with two different workarounds that I think should handle most use-cases. You can check them out [here](https://gist.github.com/jhaber/2616683c6c7c7314b2fc6ce3925d4e54)

**Comment by cowtowncoder** [comment]
Quick note: related to #209, Jackson 2.9 SHOULD allow use of:

    @JsonSetter(nulls=Nulls.AS_EMPTY)
    public String value;

to basically coerce `null` value into empty String (or one of alternatives; skip assignment to leave as default, throw exception). It's via databind feature:

https://github.com/FasterXML/jackson-databind/issues/1402

I will test that this works, and if so, mark this as closed.

**Comment by cowtowncoder** [comment]
Ok. Hmmh. So, good and bad news.

Good news: woth empty tags, works as expected.

Bad news: open+start, without even intervening space (so simple `<value></value>`) does NOT work as expected. Can not remember if (... and, how?) there was intentional checking to effect different handling. But XML semantics should not differ between two representations.

**Comment by cowtowncoder** [comment]
So. Works as suggested with empty tag, in 2.9. And with `ACCEPT_EMPTY_STRING_AS_NULL_OBJECT` can make other cases work too.

In 3.0.0 will further work the same way for "split" empty tag (start+end).

**Comment by cowtowncoder** [comment]
Note: some of the fixes to `jackson-databind` will go in 2.9.6. So above comments reflect to that upcoming version.

**Comment by itisvamsi** [comment]
Is there any fix for this which is available?
